packages/fluentui/react-northstar/src/components/Tree/Tree.tsx

```diff
@@ -96,335 +112,356 @@ export interface TreeItemForRenderProps {
+      if (selectedItemIds.indexOf(treeItemProps.id) === -1) {
+        nextSelectedItemIds = [...selectedItemIds, treeItemProps.id];
       } else {
-        const selectItems = items => {
-          items.forEach(item => {
-            const selectble = item.hasOwnProperty('selectable') ? item.selectable : treeItemProps.selectable;
-            if (selectedItemIds.indexOf(item.id) === -1) {
-              if (item.items) {
-                selectItems(item.items);
-              } else if (selectble) {
-                selectedItemIds.push(item.id);
-              }
-            }
-          });
-        };
-        selectItems(items);
+        nextSelectedItemIds = nextSelectedItemIds.filter(itemID => itemID !== treeItemProps.id);
       }
 
-      this.setSelectedItemIds(e, selectedItemIds);
-      return;
-    }
-
-    // push/remove single tree item into selection array
-    if (selectedItemIds.indexOf(id) === -1) {
-      selectedItemIds.push(id);
-    } else {
-      selectedItemIds = selectedItemIds.filter(itemID => itemID !== id);
-    }
+      setSelectedItemIds(e, nextSelectedItemIds);
+    },
+    [setSelectedItemIds],
+  );
 
-    this.setSelectedItemIds(e, selectedItemIds);
-  };
+  const expandItems = React.useCallback(
+    (e: React.SyntheticEvent, treeItemProps: TreeItemProps) => {
+      const { id } = treeItemProps;
+      const siblings = getSiblings(stableProps.current.items, id);
 
-  onTitleClick = (e: React.SyntheticEvent, treeItemProps: TreeItemProps, executeSelection: boolean = false) => {
-    const treeItemHasSubtree = hasSubtree(treeItemProps);
+      const activeItemIdIndex = activeItemIds.indexOf(id);
+      let nextActiveItemsIds = activeItemIds;
 
-    if (!treeItemProps) {
-      return;
-    }
+      if (activeItemIdIndex > -1) {
+        nextActiveItemsIds = removeItemAtIndex(activeItemIds, activeItemIdIndex);
+      } else {
+        if (exclusive) {
+          siblings.some(sibling => {
+            const activeSiblingIdIndex = activeItemIds.indexOf(sibling['id']);
+            if (activeSiblingIdIndex > -1) {
+              nextActiveItemsIds = removeItemAtIndex(activeItemIds, activeSiblingIdIndex);
 
-    if (treeItemProps.selectable) {
-      this.processItemsForSelection(e, treeItemProps, executeSelection);
-    }
+              return true;
+            }
+            return false;
+          });
+        }
 
-    if (treeItemHasSubtree && !executeSelection && e.target === e.currentTarget) {
-      this.expandItems(e, treeItemProps);
-    }
-  };
+        nextActiveItemsIds = [...nextActiveItemsIds, id];
+      }
 
-  expandItems(e: React.SyntheticEvent, treeItemProps: TreeItemProps) {
-    let { activeItemIds } = this.state;
-    const { id } = treeItemProps;
-    const { exclusive, items } = this.props;
-    const siblings = getSiblings(items, id);
+      setActiveItemIds(e, nextActiveItemsIds);
+    },
+    [stableProps, setActiveItemIds, activeItemIds.length],
```

Comment:
@layershifter I had to add this `activeItemsIds.length` otherwise the nested subtrees were not being expanded

Reply:
This is not cool 😯 Let's try to refactor it on Monday

Reply:
Kind of makes sense, `activeItemIds` is a dep and it's changing after each expand

Reply:
Do you see any other way to solve this?